File: ios/localasset-core/Sources/LocalAssetCore/api/AsyncResourceLoader.swift
import Foundation

/// A `TypedResourceLoader` whose data source is intrinsically async — e.g. a network pre-fetch
/// cache, a Room/CoreData query, or any IO that is naturally `async`.
///
/// The engine pipeline is synchronous (WebView bridges are synchronous: `WKURLSchemeHandler.start`
/// expects `didReceive`/`didFinish` to be called, and runs on a dedicated background queue). This
/// interface bridges that gap additively and non-breakingly: implement only `canLoad` and
/// `loadAsync`, and the inherited `load` runs `loadAsync` via a `Task` + `DispatchSemaphore` bridge
/// (design.md §9.3 — the documented workaround, now a first-class, typed extension point instead
/// of per-callers bridge boilerplate).
///
/// Caveat: `load` blocks the calling thread until `loadAsync` completes. The default engine call
/// site is `WKURLSchemeHandler.start`'s dedicated queue (not the Swift cooperative pool), so the
/// `Task` has pool threads to run on and the wait will not self-deadlock. Do NOT call `load` from a
/// Swift cooperative-pool thread (e.g. a `Task` body) — that can deadlock. For long sources,
/// prefer a sync loader that owns its own queue/thread.
public protocol AsyncResourceLoader: TypedResourceLoader {
    func loadAsync(descriptor: ResourceDescriptor) async throws -> ResourceData
}

public extension AsyncResourceLoader {
    /// Sync bridge over `loadAsync`. Blocks the calling thread until the spawned `Task` finishes.
    func load(descriptor: ResourceDescriptor) throws -> ResourceData {
        let semaphore = DispatchSemaphore(value: 0)
        var captured: Result<ResourceData, Error>?
        Task {
            do {
                captured = .success(try await loadAsync(descriptor: descriptor))
            } catch {
                captured = .failure(error)
            }
            semaphore.signal()
        }
        semaphore.wait()
        return try captured!.get()
    }
}

File: android/localasset-core/src/main/kotlin/io/github/xesam/android/localasset/core/api/AsyncResourceLoader.kt
package io.github.xesam.android.localasset.core.api

import io.github.xesam.android.localasset.core.model.ResourceData
import io.github.xesam.android.localasset.core.model.ResourceDescriptor
import kotlinx.coroutines.runBlocking

/**
 * A [TypedResourceLoader] whose data source is intrinsically suspending — e.g. a network pre-fetch
 * cache, a Room/DataStore query, or any IO that is naturally coroutine-based.
 *
 * The engine pipeline is synchronous (WebView bridges are synchronous: Android
 * `shouldInterceptRequest` returns a `WebResourceResponse` inline; the loader runs on a background
 * thread already). This interface bridges that gap additively and non-breakingly: implement only
 * [canLoad] and [loadAsync], and the inherited [loadTyped] runs [loadAsync] under [runBlocking] so
 * the sync pipeline can drive an async source (design.md §9.3 — the documented workaround, now a
 * first-class, typed extension point instead of per-callers `runBlocking` boilerplate).
 *
 * Caveat: [runBlocking] blocks the calling (background) thread until [loadAsync] completes. Keep
 * [loadAsync] free of main-thread hops; for truly long sources, prefer a sync loader that owns its
 * own thread/executor. Do NOT call from the main thread.
 */
public interface AsyncResourceLoader : TypedResourceLoader {
    suspend fun loadAsync(descriptor: ResourceDescriptor): ResourceData

    override fun loadTyped(descriptor: ResourceDescriptor): ResourceData = runBlocking { loadAsync(descriptor) }
}
